### PR TITLE
Change to handle external diskcache and adapted cache size limit care - issue 898

### PR DESCRIPTION
**Pull request by eldk**
Hello,

I made those changes to handle this bug : https://github.com/facebook/fresco/issues/898
+ add isExternal() boolean to check if cache is external or internal.
+ modified updateFileCacheSizeLimit() to handle internal (default) or external diskcache

Hope it helps,
Thanks,
Eric

### CODE REVIEW

**Comment by eldk** [comment]
https://github.com/facebook/fresco/issues/995#issuecomment-214986323

**Comment by facebook-github-bot-2** [comment]
@eldk updated the pull request.

**Comment by facebook-github-bot-1** [comment]
@eldk updated the pull request.

**Comment by facebook-github-bot-2** [comment]
@eldk updated the pull request.

**Comment by facebook-github-bot-8** [comment]
@eldk updated the pull request.

**Comment by facebook-github-bot-7** [comment]
@eldk updated the pull request.

**Comment by facebook-github-bot-3** [comment]
@eldk updated the pull request.

**Comment by facebook-github-bot-7** [comment]
@eldk updated the pull request.

**Comment by plamenko** [comment]
Hi! Thanks for the pull request and apologies for the delay. We will review this today.

File: imagepipeline-base/src/main/java/com/facebook/cache/disk/DefaultDiskStorage.java [hunk 1]
```diff
@@ -121,6 +124,24 @@ static String getVersionSubdirectoryName(int version) {
   public boolean isEnabled() {
     return true;
   }
+  
+  @Override
+  public boolean isExternal() {
+    boolean state = false;
```
Comment (plamenko) [inline imagepipeline-base/src/main/java/com/facebook/cache/disk/DefaultDiskStorage.java]: We could compute this only once in the constructor right after setting `mRootDirectory`.

```
mRootDirectory = rootDirectory;
mIsExternal = isExternal(rootDirectory);
```
```
private static boolean isExternal(File directory) {
  ...
}
```

File: imagepipeline-base/src/main/java/com/facebook/cache/disk/DiskStorageCache.java [hunk 2]
```diff
@@ -155,7 +156,11 @@ public DiskStorageCache(
     this.mCacheSizeLimit = params.mDefaultCacheSizeLimit;
     this.mStatFsHelper = StatFsHelper.getInstance();
 
+
     this.mStorage = diskStorage;
+
+    this.mDiskCacheIsExternal = mStorage.isExternal();
```
Comment (plamenko) [inline imagepipeline-base/src/main/java/com/facebook/cache/disk/DiskStorageCache.java]: Not needed.

File: imagepipeline-base/src/main/java/com/facebook/cache/disk/DiskStorageCache.java [hunk 3]
```diff
@@ -69,6 +69,7 @@
 
   @GuardedBy("mLock")
   private long mCacheSizeLastUpdateTime;
+  private boolean mDiskCacheIsExternal = false;
```
Comment (plamenko) [inline imagepipeline-base/src/main/java/com/facebook/cache/disk/DiskStorageCache.java]: Not needed

File: imagepipeline-base/src/main/java/com/facebook/cache/disk/DiskStorageCache.java [hunk 4]
```diff
@@ -510,10 +520,18 @@ private void evictAboveSize(
   @GuardedBy("mLock")
   private void updateFileCacheSizeLimit() {
     // Test if mCacheSizeLimit can be set to the high limit
-    boolean isAvailableSpaceLowerThanHighLimit =
+    boolean isAvailableSpaceLowerThanHighLimit;
+    if(mDiskCacheIsExternal){
+    isAvailableSpaceLowerThanHighLimit =
```
Comment (plamenko) [inline imagepipeline-base/src/main/java/com/facebook/cache/disk/DiskStorageCache.java]: No need for code duplication. We can do something like:
```
StorageType storageType = mStorage.isExternal() ? StatFsHelper.StorageType.EXTERNAL : StatFsHelper.StorageType.INTERNAL;
```

File: imagepipeline-base/src/main/java/com/facebook/cache/disk/DynamicDefaultDiskStorage.java [hunk 5]
```diff
@@ -71,6 +71,15 @@ public boolean isEnabled() {
       return false;
     }
   }
+ 
+ @Override
+  public boolean isExternal() {
+    try {
+      return get().isExternal();
```
Comment (plamenko) [inline imagepipeline-base/src/main/java/com/facebook/cache/disk/DynamicDefaultDiskStorage.java]: I don't think `isExternal` can throw `IOException` as it's already handled there.

**Comment by eldk** [comment]
Hello,
Thanks @plamenko for the review and the advices, I will check in the next days and update the commits.
Eric

**Comment by facebook-github-bot-1** [comment]
@eldk updated the pull request.

**Comment by tyronen** [comment]
It looks like you still need to incorporated @plamenko's suggestions.

**Comment by eldk** [comment]
Hello,

I've switched to Fresco 0.10.0  and testing (within my app and library modification).

There's a problem with `getMainFileCache().getCount()` 

It was working with Fresco 0.9.0 (patched) and `getMainDiskStorageCache().getCount()

I'm looking at it.

Eric

**Comment by facebook-github-bot-2** [comment]
@eldk updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@eldk updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@eldk updated the pull request.

File: imagepipeline-base/src/main/java/com/facebook/cache/disk/DynamicDefaultDiskStorage.java (hunk 5 above)
Comment (eldk) [inline imagepipeline-base/src/main/java/com/facebook/cache/disk/DynamicDefaultDiskStorage.java]: Hello,
@plamenko, here I think it's get() that throws IOException : 
` @VisibleForTesting
  /* package protected */ synchronized DiskStorage get() throws IOException {
...
}`
?

**Comment by tyronen** [comment]
@facebook-github-bot import

**Comment by facebook-github-bot-3** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/1105627192801288/int_phab) to review.

**Comment by facebook-github-bot-6** [comment]
@eldk updated the pull request.

File: imagepipeline-base/src/main/java/com/facebook/cache/disk/DefaultDiskStorage.java [hunk 6]
```diff
@@ -107,6 +116,25 @@ public DefaultDiskStorage(
     mClock = SystemClock.get();
   }
 
+  private static boolean isExternal(File directory) {
+
+    boolean state = false;
+    String cacheDirPath = null;
+    File extStoragePath = Environment.getExternalStorageDirectory();
+    if(extStoragePath != null) {
```
Comment (plamenko) [inline imagepipeline-base/src/main/java/com/facebook/cache/disk/DefaultDiskStorage.java]: Nit: coding style, add spaces around keywords

File: imagepipeline-base/src/main/java/com/facebook/cache/disk/DefaultDiskStorage.java [hunk 7]
```diff
@@ -107,6 +116,25 @@ public DefaultDiskStorage(
...
+      cacheDirPath = extStoragePath.toString();
+      try {
+        String appCacheDirPath = directory.getCanonicalPath();
+        if (appCacheDirPath.contains(cacheDirPath)) state = true;
```
Comment (plamenko) [inline imagepipeline-base/src/main/java/com/facebook/cache/disk/DefaultDiskStorage.java]: Nit: coding style, we always surround code blocks with `{ }` even if it's a single statement.

File: imagepipeline-base/src/main/java/com/facebook/cache/disk/DefaultDiskStorage.java [hunk 8]
```diff
@@ -107,6 +116,25 @@ public DefaultDiskStorage(
...
+      } catch (IOException e) {
+        e.printStackTrace();
+      }
+    }else{
```
Comment (plamenko) [inline imagepipeline-base/src/main/java/com/facebook/cache/disk/DefaultDiskStorage.java]: Nit again :)
```
} else {
  state = false;
}
```

File: imagepipeline-base/src/main/java/com/facebook/cache/disk/DiskStorageCache.java [hunk 9]
```diff
@@ -509,21 +516,33 @@ private void evictAboveSize(
...
-            StatFsHelper.StorageType.INTERNAL,
+            storageType,
             mDefaultCacheSizeLimit - mCacheStats.getSize());
     if (isAvailableSpaceLowerThanHighLimit) {
       mCacheSizeLimit = mLowDiskSpaceCacheSizeLimit;
     } else {
       mCacheSizeLimit = mDefaultCacheSizeLimit;
     }
-  }
+  }
```
Comment (plamenko) [inline imagepipeline-base/src/main/java/com/facebook/cache/disk/DiskStorageCache.java]: Nit remove trailing spaces

File: imagepipeline-base/src/main/java/com/facebook/cache/disk/DynamicDefaultDiskStorage.java (hunk 5 above)
Comment (plamenko) [inline imagepipeline-base/src/main/java/com/facebook/cache/disk/DynamicDefaultDiskStorage.java]: Okay

**Comment by plamenko** [comment]
Looks good, there are just a couple of nits related to coding conventions we employ in Fresco.

**Comment by facebook-github-bot-2** [comment]
@eldk updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@eldk updated the pull request.

**Comment by facebook-github-bot-5** [comment]
@eldk updated the pull request.

**Comment by facebook-github-bot-6** [comment]
@eldk updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@eldk updated the pull request.

**Comment by kirwan** [comment]
@facebook-github-bot import

**Comment by facebook-github-bot-8** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/1105627192801288/int_phab) to review internal test results.

File: imagepipeline-base/src/main/java/com/facebook/cache/disk/DiskStorageCache.java [hunk 10]
```diff
@@ -593,8 +601,18 @@ private void updateFileCacheSizeLimit() {
...
+        }
     return mCacheStats.getSize();
   }
+  
+  public long getCount() {
+    if (!mCacheStats.isInitialized()){
+            maybeUpdateFileCacheSize();
+        }
+    return mCacheStats.getCount();
+  }
```
Comment (kirwan) [inline imagepipeline-base/src/main/java/com/facebook/cache/disk/DiskStorageCache.java]: When testing internally, I've found a couple of issues here.

`maybeUpdateFileCacheSize()` should be guarded by `synchronized(mLock)`.

In any case though, I don't think you should need to call that in either of these methods now as we've recently added an in-memory cache and the stats are now initialized from the constructor.

**Comment by eldk** [comment]
Ok, I will have a look.

**Comment by lambdapioneer** [comment]
@facebook-github-bot import

**Comment by facebook-github-bot-1** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/1105627192801288/int_phab) to review internal test results.

**Comment by lambdapioneer** [comment]
Hi @eldk, thanks a lot for the diff. I'm currently testing and merging your PR.

One quick question: Is there a particular reason for having `isExternal` in `FileCache`? Seems to me that we can just have it in `DiskStorage`. There might be FileCaches where this method cannot provide a valuable answer: For example, if the cache is on both internal and external storage.

If you are fine with the change, I'd go ahead and remove `isExternal` from the `FileCache` interface. That should not change the benefit from this PR

**Comment by eldk** [comment]
Hello,
>If you are fine with the change, I'd go ahead and remove isExternal from the FileCache interface. 
It's OK.
Thanks,
Eric

Thanks,

Eric

**Pull request closed without merging**